core/src/main/java/org/apache/iceberg/metrics/TimerResultParser.java

```diff
@@ -0,0 +1,125 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.metrics;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.TimeUnit;
+import org.apache.iceberg.metrics.ScanReport.TimerResult;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.JsonUtil;
+
+class TimerResultParser {
+  private static final String MISSING_FIELD_ERROR_MSG =
+      "Cannot parse timer from '%s': Missing field '%s'";
+
+  private static final String NAME = "name";
+  private static final String TIME_UNIT = "time-unit";
+  private static final String COUNT = "count";
+  private static final String TOTAL_DURATION = "total-duration";
+
+  private TimerResultParser() {}
+
+  static String toJson(TimerResult timerResult) {
+    return toJson(timerResult, false);
+  }
+
+  static String toJson(TimerResult timerResult, boolean pretty) {
+    return JsonUtil.generate(gen -> toJson(timerResult, gen), pretty);
+  }
+
+  static void toJson(TimerResult timer, JsonGenerator gen) throws IOException {
+    Preconditions.checkArgument(null != timer, "Invalid timer: null");
+
+    gen.writeStartObject();
+    gen.writeStringField(NAME, timer.name());
+    gen.writeNumberField(COUNT, timer.count());
+    gen.writeStringField(TIME_UNIT, timer.timeUnit().name());
+    gen.writeNumberField(TOTAL_DURATION, fromDuration(timer.totalDuration(), timer.timeUnit()));
+    gen.writeEndObject();
+  }
+
+  static TimerResult fromJson(String json) {
+    return JsonUtil.parse(json, TimerResultParser::fromJson);
+  }
+
+  static TimerResult fromJson(String property, JsonNode json) {
+    return fromJson(get(property, json));
+  }
+
+  static TimerResult fromJson(JsonNode json) {
+    Preconditions.checkArgument(null != json, "Cannot parse timer from null object");
+    Preconditions.checkArgument(json.isObject(), "Cannot parse timer from non-object: %s", json);
+
+    String name = JsonUtil.getString(NAME, json);
+    long count = JsonUtil.getLong(COUNT, json);
+    TimeUnit unit = TimeUnit.valueOf(JsonUtil.getString(TIME_UNIT, json));
+    long duration = JsonUtil.getLong(TOTAL_DURATION, json);
+    return new TimerResult(name, unit, toDuration(duration, unit), count);
+  }
+
+  private static JsonNode get(String property, JsonNode node) {
```

Comment:
This method name doesn't describe what's happening here. This is validating that the node is a timer. Maybe this should be `validate` instead?

Reply:
the only reason this method and the same one in the `CounterResultParser` exists is shown in the two tests below
```
  @Test
  public void invalidTimer() {
    Assertions.assertThatThrownBy(
            () ->
                ScanMetricsResultParser.fromJson(
                    "{\"total-planning-duration\":{\"name\":\"resultDataFiles\",\"unit\":\"COUNT\",\"value\":5}}"))
        .isInstanceOf(IllegalArgumentException.class)
        .hasMessage("Cannot parse timer from 'total-planning-duration': Missing field 'count'");
  }

  @Test
  public void invalidCounter() {
    Assertions.assertThatThrownBy(
            () ->
                ScanMetricsResultParser.fromJson(
                    "{\"total-planning-duration\":{\"name\":\"totalPlanningDuration\",\"count\":1,\"time-unit\":\"NANOSECONDS\",\"total-duration\":600000000000},"
                        + "\"result-data-files\":{\"name\":\"totalPlanningDuration\",\"count\":1,\"time-unit\":\"NANOSECONDS\",\"total-duration\":600000000000}}"))
        .isInstanceOf(IllegalArgumentException.class)
        .hasMessage("Cannot parse counter from 'result-data-files': Missing field 'unit'");
  }
```

Just using `JsonUtil` directly here would only tell us that the timer is missing field X but you don't know whether that field was missing in `total-planning-duration` / `result-data-files` or any other field.

Also note the error message `Cannot parse counter from 'result-data-files': Missing field 'unit'` using single quotes, which doesn't necessarily conform to how we do error messaging. So I was wondering whether you have a good suggestion on how to represent this error message in the format we typically use?

Reply:
I think it's a good idea to have this method and provide context. But I think that the name should describe what it's doing a bit better.

For the error message, I would just omit the quotes and move the colon to just before the missing field (so that it is clear something is intended to come after). If one or both are empty/null then you get a reasonable error message that I think most people would be able to interpret as being built from empty strings:

```
Cannot parse counter from , missing field: 
```